kernel/posix/pthread_rwlock.c

```diff
@@ -0,0 +1,304 @@
+		return EINVAL;
+	}
+
+	timeout = calculate_timeout(abstime);
+
+	if (k_sem_take(&rwlock->sem, timeout) == 0) {
+		rwlock->owner = k_current_get();
+	} else {
+		ret = ETIMEDOUT;
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Lock a read-write lock object for writing immedately.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_rwlock_trywrlock(pthread_rwlock_t *rwlock)
+{
+	if (rwlock->status == NOT_INITIALIZED) {
+		return EINVAL;
+	}
+
+	if (k_sem_take(&rwlock->sem, K_NO_WAIT) == 0) {
+		rwlock->owner = k_current_get();
+		return 0;
+	}
+
+	return EBUSY;
+}
+
+/**
+ *
+ * @brief Unlock a read-write lock object.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_rwlock_unlock(pthread_rwlock_t *rwlock)
+{
+	s32_t count;
+
+	if (rwlock->status == NOT_INITIALIZED) {
+		return EINVAL;
+	}
+
+	k_sem_take(&rwlock->count_lock, K_FOREVER);
+
+	if (rwlock->current_reader_count != 0) {
+		/*  Read unlock */
+		rwlock->current_reader_count -= 1;
+
+		if (rwlock->current_reader_count == 0) {
+			rwlock->owner = NULL;
+			k_sem_give(&rwlock->sem);
+		}
+	} else {
+		/* Write unlock */
+		k_sem_give(&rwlock->sem);
+
+		/* Unblock all readers */
+		for (count = 0; count < rwlock->blocked_reader_count; count++) {
+			k_sem_give(&rwlock->blocking_sem);
+		}
+
+		rwlock->owner = NULL;
+	}
+
+	k_sem_give(&rwlock->count_lock);
+	return 0;
+}
+
+static int read_lock_acquire(pthread_rwlock_t *rwlock, u32_t timeout)
+{
+	int ret = 0;
+
+	k_sem_take(&rwlock->count_lock, K_FOREVER);
+
+	if (rwlock->current_reader_count > 0) {
+		rwlock->current_reader_count += 1;
+		k_sem_give(&rwlock->count_lock);
+		return ret;
+	}
+
+	if (k_sem_take(&rwlock->sem, K_NO_WAIT) == 0) {
+		rwlock->current_reader_count += 1;
+		rwlock->owner = k_current_get();
+		k_sem_give(&rwlock->count_lock);
+		return ret;
+	}
+
+	k_sem_give(&rwlock->count_lock);
+
+	if (timeout == K_NO_WAIT) {
+		return EBUSY;
+	}
+
+	rwlock->blocked_reader_count += 1;
+
+	/*
+	 * A reader thread may be blocked multiple times if a writer thread
+	 * acquires lock.
+	 */
+	for (;;) {
+		if (k_sem_take(&rwlock->blocking_sem, timeout) != 0) {
+			k_sem_take(&rwlock->count_lock, K_FOREVER);
+			rwlock->blocked_reader_count -= 1;
+			ret = ETIMEDOUT;
+			break;
+		}
+
+		k_sem_take(&rwlock->count_lock, K_FOREVER);
+
+		if (rwlock->current_reader_count > 0) {
+			rwlock->current_reader_count += 1;
+			rwlock->blocked_reader_count -= 1;
+			break;
+		}
+
+		/*
+		 * Try again to take rwlock->sem, since another writer or reader
+		 * may have taken it in the meantime.
+		 */
+		if (k_sem_take(&rwlock->sem, 0) == 0) {
+			rwlock->current_reader_count += 1;
+			rwlock->blocked_reader_count -= 1;
+			rwlock->owner = k_current_get();
+			break;
+		}
+
+		k_sem_give(&rwlock->count_lock);
+
+	}
+	k_sem_give(&rwlock->count_lock);
+
+	return ret;
+}
+
+static u32_t calculate_timeout(const struct timespec *abstime)
+{
+	s64_t milli_secs;
+	s32_t secs;
+	struct timespec curtime;
+	u32_t timeout;
+
+	clock_gettime(CLOCK_MONOTONIC, &curtime);
```

Comment:
Are we using CLOCK_MONOTONIC here due to the lack of CLOCK_REALTIME?

Reply:
Yes we are using monotonic clock. We do not have CLOCK_REALTIME.